src/test.js

```diff
@@ -692,47 +696,112 @@ export function test( testName, callback ) {
 	newTest.queue();
 }
 
-extend( test, {
-	todo: function todo( testName, callback ) {
-		if ( focused || config.currentModule.ignored ) {
-			return;
-		}
+function todo( testName, data, callback ) {
+	if ( focused || config.currentModule.ignored ) {
+		return;
+	}
 
-		const newTest = new Test( {
-			testName,
-			callback,
-			todo: true
-		} );
+	const newTest = new Test( {
+		testName,
+		callback,
+		todo: true,
+		params: data
+	} );
 
-		newTest.queue();
-	},
-	skip: function skip( testName ) {
-		if ( focused || config.currentModule.ignored ) {
-			return;
-		}
+	newTest.queue();
+}
 
-		const test = new Test( {
-			testName: testName,
-			skip: true
-		} );
+function skip( testName ) {
+	if ( focused || config.currentModule.ignored ) {
+		return;
+	}
+
+	const test = new Test( {
+		testName: testName,
+		skip: true
+	} );
+
+	test.queue();
+}
+function only( testName, data, callback ) {
+	if ( config.currentModule.ignored ) {
+		return;
+	}
+	if ( !focused ) {
+		config.queue.length = 0;
+		focused = true;
+	}
 
-		test.queue();
+	const newTest = new Test( {
+		testName: testName,
+		callback: callback,
+		params: data
+	} );
+
+	newTest.queue();
+}
+
+function normalizeTable( data ) {
+	const isTable = data.every( Array.isArray );
+	if ( isTable ) {
+		return data;
+	}
+	return data.map( datum => [ datum ] );
+}
+
+function runEach( data, eachFn ) {
+	if ( Array.isArray( data ) ) {
+		normalizeTable( data ).forEach( eachFn );
+	} else {
+		throw new Error(
+			`test.each expects array of arrays or an array of primitives is the expected input.
+${typeof data} was found instead.`
+		);
+	}
+}
+
+extend( test, {
+	todo: function( testName, callback ) {
+		todo( testName, undefined, callback );
+	},
+	skip: function( testName ) {
+		skip( testName );
+	},
+	only: function( testName, callback ) {
+		only( testName, undefined, callback );
 	},
-	only: function only( testName, callback ) {
+	each: function( testName, data, callback ) {
 		if ( config.currentModule.ignored ) {
 			return;
 		}
```

Comment:
what about if `focused`, like the `test` implementation?

Is there anyway to reuse that logic? Could we have a local `addTestWithData` function that both `test` and `each` leverage? Right now we could make `test` really call `each` to reuse logic, but that smells funny, and I think highlights the need for a lower level helper function.

If this is creeping scope into "pretty code for pretty sake", then I'm okay with that; just looking for where to lean down to reduce edge cases that have already been solved before.

Reply:
That makes a lot of sense. Refactored the code into `addTestWithData`.